### PR TITLE
Test.properties 1339

### PR DESCRIPTION
**Pull request by vladb00mer**
!see commit comments!

### CODE REVIEW

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

File: jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+public class DriversVersionTests {
+
+    @DataProvider
+    public static Object[] negativeData() {
+        return new Object[]{"latest", "LATES", "prelatest", "", "RELATEST", "2,31", "ten.eleven"};
+    }
+
+    @DataProvider
+    public static Object[] positiveData() {
+        return new Object[]{"LATEST", "1.23", "PRELATEST", "33.44"};
```
Comment (splekhanov) [inline jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java]: Lets see valid values for this case: https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-java
1. The third number always exists. Even if it's zero, so 1.23 and 33.44 are invalid, lets get rid of them in this test
2. Lets add some valid cases: 
- 4.0.0-alpha-3
- 3.141.59

File: jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import static com.epam.jdi.light.common.Property.DRIVERS_VERSION;
+import static com.epam.jdi.light.common.PropertyValidationUtils.validateProperties;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class DriversVersionTests {
+
+    @DataProvider
+    public static Object[] negativeData() {
+        return new Object[]{"latest", "LATES", "prelatest", "", "RELATEST", "2,31", "ten.eleven"};
```
Comment (splekhanov) [inline jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java]: Lets add some invalid data:
- 2.33
- 2.34.
- 1.4.*

File: jdi-light/src/main/java/com/epam/jdi/light/common/Property.java [hunk 3]
```diff
@@ -3,7 +3,13 @@
 public enum Property {
 
     TIMEOUT_WAIT_ELEMENT("timeout.wait.element", "^[1-9][0-9]{1,2}$|^\\d$",
-            "Value of 'timeout.wait.element' must be an integer from 0 to 999.");
+            "Value of 'timeout.wait.element' must be an integer from 0 to 999."),
+
+    DRIVER("driver", "^\\$\\{driver\\}$|^chrome$|^firefox$|^ie$", "driver parameter value is incorrect"),
+
+    DRIVERS_VERSION("drivers.version", "LATEST|PRELATEST|[0-9]+[.][0-9][0-9]", "drivers.version parameter value is incorrect"),
```
[the remark this answers is not in the document]
Comment (splekhanov) [reply jdi-light/src/main/java/com/epam/jdi/light/common/Property.java]: It seems like Java catches those matches if we use just bare words for a strict match.
Lets use "LATEST|PRELATEST|^([0-9])+([.][0-9]{1,3}[.][a-zA-Z0-9_-]+|)$"
and "^\\$\\{driver\\}$|chrome|firefox|ie"

[the remark this answers is not in the document]
Comment (vladb00mer) [reply jdi-light/src/main/java/com/epam/jdi/light/common/Property.java]: updated but errors in tests

File: jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java (hunk 1 above)
Comment (vladb00mer) [reply jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java]: test data actualized, all tests are passed

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

File: jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java (hunk 2 above)
Comment (vladb00mer) [reply jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java]: added

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

**Comment by codacy-review** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- jdi-light/src/test/java/io/github/epam/properties/DriversVersionTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriverTests.java  2
- jdi-light/src/test/java/io/github/epam/properties/DriversFolderTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/jdi-light/jdi-light/pullRequest?prid=4576573&bid=15377733)

**Pull request merged**